pytensor/tensor/rewriting/math.py

```diff
@@ -167,6 +170,65 @@ def local_0_dot_x(fgraph, node):
             return [constant_zero]
 
 
+@register_canonicalize
+@register_specialize
+@register_stabilize
+@node_rewriter([Dot])
+def local_block_diag_dot_to_dot_block_diag(fgraph, node):
+    r"""
+    Perform the rewrite ``dot(block_diag(A, B), C) -> concat(dot(A, C), dot(B, C))``
+
+    BlockDiag results in the creation of a matrix of shape ``(n1 * n2, m1 * m2)``. Because dot has complexity
+    of approximately O(n^3), it's always better to perform two dot products on the smaller matrices, rather than
+    a single dot on the larger matrix.
+    """
+    x, y = node.inputs
+    op = node.op
+
+    def check_for_block_diag(x):
```

Comment:
Since this is likely very rare compared to Dot, may it be worth to track the rewrite on BlockDiagonal instead? Like we did in https://github.com/pymc-devs/pytensor/pull/1483/files with `expm1` and the use `clients_at_depth` and use a return dictionary

Reply:
Yes, good idea. I didn't have a good sense of the right thing to track. This is also why it doesn't track BlockwiseDot